In [2]:
#  Step 1: Install Required Tools
!apt-get update
!apt-get install -y sra-toolkit bwa samtools bcftools fastqc

Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]
Get:2 https://cli.github.com/packages stable InRelease [3,917 B]
Get:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1,581 B]
Hit:4 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:5 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
Get:6 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]
Get:7 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [2,103 kB]
Get:8 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]
Hit:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Hit:10 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Hit:11 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease
Get:12 https://cli.github.com/packages stable/main amd64 Packages [344 B]
Get:13 http://archive.ubuntu.com/ubuntu jammy-backports InReleas

In [4]:
# Step 2: Download FASTQ Data for SRR9200893
!prefetch SRR9200893
!fastq-dump --split-files SRR9200893


2025-11-06T05:30:24 prefetch.2.11.3: Current preference is set to retrieve SRA Normalized Format files with full base quality scores.
2025-11-06T05:30:24 prefetch.2.11.3: 1) Downloading 'SRR9200893'...
2025-11-06T05:30:24 prefetch.2.11.3: SRA Normalized Format file is being retrieved, if this is different from your preference, it may be due to current file availability.
2025-11-06T05:30:24 prefetch.2.11.3:  Downloading via HTTPS...
2025-11-06T05:30:33 prefetch.2.11.3:  HTTPS download succeed
2025-11-06T05:30:35 prefetch.2.11.3:  'SRR9200893' is valid
2025-11-06T05:30:35 prefetch.2.11.3: 1) 'SRR9200893' was downloaded successfully
Read 2000000 spots for SRR9200893
Written 2000000 spots for SRR9200893


In [6]:
# Step 3: Run FastQC
!fastqc SRR9200893_1.fastq -o /content/
!fastqc SRR9200893_2.fastq -o /content/

Started analysis of SRR9200893_1.fastq
Approx 5% complete for SRR9200893_1.fastq
Approx 10% complete for SRR9200893_1.fastq
Approx 15% complete for SRR9200893_1.fastq
Approx 20% complete for SRR9200893_1.fastq
Approx 25% complete for SRR9200893_1.fastq
Approx 30% complete for SRR9200893_1.fastq
Approx 35% complete for SRR9200893_1.fastq
Approx 40% complete for SRR9200893_1.fastq
Approx 45% complete for SRR9200893_1.fastq
Approx 50% complete for SRR9200893_1.fastq
Approx 55% complete for SRR9200893_1.fastq
Approx 60% complete for SRR9200893_1.fastq
Approx 65% complete for SRR9200893_1.fastq
Approx 70% complete for SRR9200893_1.fastq
Approx 75% complete for SRR9200893_1.fastq
Approx 80% complete for SRR9200893_1.fastq
Approx 85% complete for SRR9200893_1.fastq
Approx 90% complete for SRR9200893_1.fastq
Approx 95% complete for SRR9200893_1.fastq
Approx 100% complete for SRR9200893_1.fastq
Analysis complete for SRR9200893_1.fastq
Started analysis of SRR9200893_2.fastq
Approx 5% complete fo

In [8]:
# Step 4: Download and Index Human Reference Genome (GRCh38)
!wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/GRCh38.primary_assembly.genome.fa.gz
!gunzip GRCh38.primary_assembly.genome.fa.gz
!bwa index GRCh38.primary_assembly.genome.fa

--2025-11-06 05:33:36--  ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/GRCh38.primary_assembly.genome.fa.gz
           => ‘GRCh38.primary_assembly.genome.fa.gz’
Resolving ftp.ebi.ac.uk (ftp.ebi.ac.uk)... 193.62.193.165
Connecting to ftp.ebi.ac.uk (ftp.ebi.ac.uk)|193.62.193.165|:21... connected.
Error in server response. Closing.
Retrying.

--2025-11-06 05:33:37--  ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/GRCh38.primary_assembly.genome.fa.gz
  (try: 2) => ‘GRCh38.primary_assembly.genome.fa.gz’
Connecting to ftp.ebi.ac.uk (ftp.ebi.ac.uk)|193.62.193.165|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /pub/databases/gencode/Gencode_human/release_44 ... done.
==> SIZE GRCh38.primary_assembly.genome.fa.gz ... 844691642
==> PASV ... done.    ==> RETR GRCh38.primary_assembly.genome.fa.gz ... done.
Length: 844691642 (806M) (unauthoritative)

GRCh38.primary_asse 100%[

In [10]:
# Step 5: Align Reads to Genome
!bwa mem GRCh38.primary_assembly.genome.fa SRR9200893_1.fastq SRR9200893_2.fastq > aligned.sam

[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 100000 sequences (10000000 bp)...
[M::process] read 100000 sequences (10000000 bp)...
[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (20, 30565, 12, 21)
[M::mem_pestat] analyzing insert size distribution for orientation FF...
[M::mem_pestat] (25, 50, 75) percentile: (170, 221, 566)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1358)
[M::mem_pestat] mean and std.dev: (326.58, 264.39)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 1754)
[M::mem_pestat] analyzing insert size distribution for orientation FR...
[M::mem_pestat] (25, 50, 75) percentile: (126, 188, 355)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 813)
[M::mem_pestat] mean and std.dev: (198.51, 128.00)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 1042)
[M::mem_pestat] analyzing insert size distribution for orientation RF...
[M::mem_pestat] (25, 50, 75) per

In [12]:
#  Step 6: Convert SAM to BAM, Sort and Index
!samtools view -Sb aligned.sam > aligned.bam
!samtools sort aligned.bam -o sorted.bam
!samtools index sorted.bam

[bam_sort_core] merging from 1 files and 1 in-memory blocks...


In [14]:
#  Step 7: Variant Calling
!samtools mpileup -uf GRCh38.primary_assembly.genome.fa sorted.bam | bcftools call -mv -Ov -o variants.vcf

[warning] samtools mpileup option `u` is functional, but deprecated. Please switch to using bcftools mpileup in future.
Note: none of --samples-file, --ploidy or --ploidy-file given, assuming all sites are diploid
[mpileup] 1 samples in 1 input files


In [15]:
# Step 8: Preview First 20 Lines of VCF File
with open('variants.vcf') as f:
    for _ in range(20):
        print(f.readline().strip())


##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##samtoolsVersion=1.13+htslib-1.13+ds
##samtoolsCommand=samtools mpileup -uf GRCh38.primary_assembly.genome.fa sorted.bam
##reference=file://GRCh38.primary_assembly.genome.fa
##contig=<ID=chr1,length=248956422>
##contig=<ID=chr2,length=242193529>
##contig=<ID=chr3,length=198295559>
##contig=<ID=chr4,length=190214555>
##contig=<ID=chr5,length=181538259>
##contig=<ID=chr6,length=170805979>
##contig=<ID=chr7,length=159345973>
##contig=<ID=chr8,length=145138636>
##contig=<ID=chr9,length=138394717>
##contig=<ID=chr10,length=133797422>
##contig=<ID=chr11,length=135086622>
##contig=<ID=chr12,length=133275309>
##contig=<ID=chr13,length=114364328>
##contig=<ID=chr14,length=107043718>
##contig=<ID=chr15,length=101991189>


Here is a detailed, line-by-line interpretation of the findings in your `variants.vcf` file, based on the first 20 lines you previewed:

*   **`##fileformat=VCFv4.2`**: This line specifies the version of the Variant Call Format (VCF) used in this file, which is version 4.2.
*   **`##FILTER=<ID=PASS,Description="All filters passed">`**: This line defines a FILTER that indicates whether a variant has passed all quality filters. In this case, a variant with `FILTER=PASS` is considered a high-quality call.
*   **`##samtoolsVersion=1.13+htslib-1.13+ds`**: This indicates the version of Samtools used to generate part of this VCF file (specifically, the mpileup step).
*   **`##samtoolsCommand=samtools mpileup -uf GRCh38.primary_assembly.genome.fa sorted.bam`**: This line records the exact `samtools mpileup` command used, including the reference genome and the sorted BAM file.
*   **`##reference=file://GRCh38.primary_assembly.genome.fa`**: This specifies the path to the reference genome file used for alignment and variant calling.
*   **`##contig=<ID=chr1,length=248956422>`** (and similar lines for other chromosomes): These lines define the contigs (chromosomes or sequences) present in the reference genome and their lengths. This information is important for locating variants within the genome.
*   Subsequent lines (not fully shown in the preview but expected in a VCF) would represent the actual variant calls. Each of these lines typically starts with chromosome and position information and includes details about the reference allele, alternate allele(s), quality score, and other relevant information about the variant and the sample.

In summary, the initial lines of the VCF file provide metadata about the file format, the tools and commands used, the reference genome, and the chromosomes it contains. The lines that follow (not fully shown in the preview) contain the specific details of each identified genetic variation.

Here is a summary of the steps we have completed in this notebook:

1.  **Installed Required Tools**: We installed essential bioinformatics tools including sra-toolkit, bwa, samtools, bcftools, and fastqc.
2.  **Downloaded FASTQ Data**: We downloaded sequencing data for the SRA accession SRR9200893 using `prefetch` and extracted the FASTQ files using `fastq-dump`.
3.  **Ran FastQC**: We performed quality control on the downloaded FASTQ files using `fastqc` to assess the quality of the sequencing reads.
4.  **Downloaded and Indexed Human Reference Genome**: We downloaded the GRCh38 human reference genome and indexed it using `bwa index` for alignment.
5.  **Aligned Reads to Genome**: We aligned the FASTQ reads to the indexed reference genome using `bwa mem`, generating a SAM file.
6.  **Converted SAM to BAM, Sorted and Indexed**: We converted the SAM file to BAM format, sorted the BAM file, and created an index using `samtools`.
7.  **Variant Calling**: We performed variant calling using `samtools mpileup` and `bcftools call` to identify genetic variations and generated a VCF file.
8.  **Previewed VCF File**: We displayed the first few lines of the resulting VCF file to examine its format and content.
9.  **Interpreted VCF Header**: We provided a line-by-line interpretation of the header section of the VCF file.
10. **Provided Final Conclusion**: We summarized the successful completion of the basic variant calling pipeline and the significance of the VCF file.